File: game/src/corridor.cpp
#include "corridor.hpp"
#include <algorithm>

namespace dts {
void corridor::connect(sinen::grid<int> &grid, std::vector<room> &rooms,
                       float floor_corridor) {
  // rooms sort by x
  std::sort(rooms.begin(), rooms.end(), [](const room &a, const room &b) {
    return a.get_position().x < b.get_position().x;
  });

  for (int i = 0; i < rooms.size() - 1;) {
    // choice room edge
    sinen::point2i left = rooms[i].get_position();
    left.x += rooms[i].get_size().x - 1;
    left.y = sinen::random::get_int_range(left.y,
                                          left.y + rooms[i].get_size().y - 1);
    sinen::point2i right = rooms[i + 1].get_position();
    right.y = sinen::random::get_int_range(
        right.y, right.y + rooms[i + 1].get_size().y - 1);

    // connect
    sinen::point2i current = left;
    while (current.x < right.x) {
      grid[current.y][current.x] = floor_corridor;
      ++current.x;
    }

    while (current.y != right.y) {
      grid[current.y][current.x] = floor_corridor;
      if (current.y < right.y) {
        ++current.y;
      } else {
        --current.y;
      }
    }

    ++i;
  }
}
} // namespace dts

File: examples/top_down_shooting/script/dungeon_generator/corridor.lua
--------------------------------------------------------------------
-- corridor.lua
--------------------------------------------------------------------
local Corridor = {}
Corridor.__index = Corridor

function Corridor:new()
  return setmetatable({ path = {} }, Corridor)
end

function Corridor:get_path() return self.path end

function Corridor:connect(grid, rooms, floor_id)
  for i = 1, #rooms - 1 do
    local a       = rooms[i]
    local b       = rooms[i + 1]
    local posA    = a:get_position()
    local sizeA   = a:get_size()
    local posB    = b:get_position()
    local sizeB   = b:get_size()

    local left    = Point2i(
      posA.x + sizeA.x - 1,
      math.random(posA.y, posA.y + sizeA.y - 1)
    )
    local right   = Point2i(
      posB.x,
      math.random(posB.y, posB.y + sizeB.y - 1)
    )

    local current = Point2i(left.x, left.y)
    while current.x < right.x do
      grid:set(current.x, current.y, floor_id)
      table.insert(self.path, Point2i(current.x, current.y))
      current.x = current.x + 1
    end

    while current.y ~= right.y do
      grid:set(current.x, current.y, floor_id)
      if current.y < right.y then
        table.insert(self.path, Point2i(current.x, current.y))
        current.y = current.y + 1
      else
        current.y = current.y - 1
      end
    end
  end
end

return Corridor
